### PR TITLE
make creating genesis block independent of addr and sig

### PR DESCRIPTION
**Pull request by MrAngelll**

### CODE REVIEW

File: blockchain/genesis.go [hunk 1]
```diff
@@ -36,34 +35,29 @@ type Genesis struct {
...
+	PriKey string `yaml:"priKey"`
 }
 
 // Transfer is the Transfer struct
 type Transfer struct {
-	Amount    int64  `yaml:"amount"`
-	Recipient string `yaml:"recipient"`
-	Signature string `yaml:"signature"`
+	Amount       int64  `yaml:"amount"`
+	RecipientPub string `yaml:"recipientPub"`
```
Comment (zjshen14) [inline blockchain/genesis.go]: RecipientPub -> RecipientPK?

File: blockchain/genesis.go [hunk 2]
```diff
@@ -73,9 +67,37 @@ var Gen = &Genesis{
 	Timestamp:           uint64(1524676419),
 	ParentHash:          hash.Hash32B{},
 	GenesisCoinbaseData: "Connecting the physical world, block by block",
-	CreatorPubKey:       "d01164c3afe47406728d3e17861a3251dcff39e62bdc2b93ccb69a02785a175e195b5605517fd647eb7dd095b3d862dffb087f35eacf10c6859d04a100dbfb7358eeca9d5c37c904",
-	// TODO: move this out of production code, which is only used for test
-	CreatorPrivKey: "d2df3528ff384d41cc9688c354cd301a09f91d95582eb8034a6eff140e7539cb17b53401",
+}
+
+// DecodeKey decodes the string keypair
+func decodeKey(pubK string, priK string) (pk keypair.PublicKey, sk keypair.PrivateKey) {
```
Comment (zjshen14) [inline blockchain/genesis.go]: Usually we put private methods (with lowerCamelCase) after public method

File: blockchain/genesis.go [hunk 3]
```diff
@@ -73,9 +67,37 @@ var Gen = &Genesis{
...
+		privateKey, err := keypair.DecodePrivateKey(priK)
+		if err != nil {
+			logger.Panic().Err(err).Msg("Fail to decode private key")
+		}
+		sk = privateKey
+	}
+	return
+}
+
+// GenerateAddr returns the string address according to public key
```
Comment (zjshen14) [inline blockchain/genesis.go]: method name doesn't match the comment

File: blockchain/genesis.go [hunk 4]
```diff
@@ -73,9 +67,37 @@ var Gen = &Genesis{
 	Timestamp:           uint64(1524676419),
 	ParentHash:          hash.Hash32B{},
 	GenesisCoinbaseData: "Connecting the physical world, block by block",
-	CreatorPubKey:       "d01164c3afe47406728d3e17861a3251dcff39e62bdc2b93ccb69a02785a175e195b5605517fd647eb7dd095b3d862dffb087f35eacf10c6859d04a100dbfb7358eeca9d5c37c904",
-	// TODO: move this out of production code, which is only used for test
-	CreatorPrivKey: "d2df3528ff384d41cc9688c354cd301a09f91d95582eb8034a6eff140e7539cb17b53401",
+}
+
+// DecodeKey decodes the string keypair
```
Comment (zjshen14) [inline blockchain/genesis.go]: Method name doesn't match the comment (case)

File: blockchain/genesis.go [hunk 5]
```diff
@@ -95,47 +117,53 @@ func NewGenesisBlock(cfg *config.Config) *Block {
...
+	creatorPubk, creatorPrik := decodeKey(Gen.CreatorPubKey, Gen.CreatorPrivKey)
+	creatorAddr := Gen.CreatorAddr(cfg.Chain.ID)
+
 	votes := []*action.Vote{}
 	for _, nominator := range actions.SelfNominators {
-		pubk, err := keypair.DecodePublicKey(nominator.PubKey)
+		pk, sk := decodeKey(nominator.PubKey, nominator.PriKey)
+		address := generateAddr(cfg.Chain.ID, pk)
+		vote, err := action.NewVote(
+			0,
```
Comment (zjshen14) [inline blockchain/genesis.go]: @lizhefeng, would you please comment if 0 nonce is correct or not?

File: blockchain/genesis.go [hunk 6]
```diff
@@ -95,47 +117,53 @@ func NewGenesisBlock(cfg *config.Config) *Block {
...
 	for _, transfer := range actions.Transfers {
-		signature, err := hex.DecodeString(transfer.Signature)
-		if err != nil {
-			logger.Panic().Err(err).Msg("Fail to create genesis block")
-		}
+		rpk, _ := decodeKey(transfer.RecipientPub, "")
+		recipientAddr := generateAddr(cfg.Chain.ID, rpk)
 		tsf, err := action.NewTransfer(
-			0, big.NewInt(transfer.Amount), Gen.CreatorAddr(cfg.Chain.ID), transfer.Recipient, []byte{}, 0, big.NewInt(0))
+			0,
```
Comment (zjshen14) [inline blockchain/genesis.go]: @lizhefeng same here

File: blockchain/genesis.go (hunk 5 above)
Comment (lizhefeng) [reply blockchain/genesis.go]: This is correct. Meant to be 0.

File: blockchain/genesis.go (hunk 6 above)
Comment (lizhefeng) [reply blockchain/genesis.go]: This is correct. Meant to be 0.

Comment (MrAngelll) [reply blockchain/genesis.go]: I have refactored the code

**Comment by zjshen14** [comment]
@MrAngelll congrats on the first PR. Looking forward to more contributions

**Pull request merged**